supply/lib/supply/uploader.rb

```diff
@@ -81,6 +85,21 @@ def perform_upload_meta(version_codes, track_name)
           version_codes.to_s == ""
         end
 
+        upload_worker = FastlaneCore::QueueWorker.new do |job|
+          begin
+            UI.message("Preparing uploads for language '#{job.language}'...")
+            start_time = Time.now
+            listing = client.listing_for_language(job.language)
+            upload_metadata(job.language, listing) unless Supply.config[:skip_upload_metadata]
+            upload_images(job.language) unless Supply.config[:skip_upload_images]
+            upload_screenshots(job.language) unless Supply.config[:skip_upload_screenshots]
+            job.release_notes << upload_changelog(job.language, job.version_code) unless Supply.config[:skip_upload_changelogs]
```

Comment:
I am not 100% sure if this really works well in a multi-threaded environment? Especially since Ruby mostly works with values, not references like for the concept of `inout` parameter in some other languages.

Basically you initialize local variable `release_notes = []` on line 110, then pass the _value_ of that var when building your `UploadJob` instance that you enqueue. Then for each job you append a new value to `job.release_notes`… but is that the same reference as the initial `release_notes` variable you created on line 110, as opposed to be a passed-by-value copy?? 🤔 

(Also, even if it was passed by reference, what happens if multiple `UploadJobs` are finishing potentially out of order and try to append to that array concurrently? 🤔 💭 )

---

_Maybe_ this works in practice though, because I _think_ that in some way ruby Arrays "act as references"? So adding a value to an array you passed as parameter will add to the original (not to a copied of the passed-by-value array)… 🤔  But even if it does, I'm not sure it feels natural to rely on this.

Instead, wdyt about:
 - Removing `:release_notes` field from your `UploadJob` struct
 - Declaring `release_notes = []` before you declare your `upload_worker` and its block (so that the variable is declared and can be captured by the block declaration)
 - Inside your block, use `release_notes << upload_changelog(…)` to directly append to the outside, captured variable

This approach would be very similar to [the example that is given in the official doc of `Thread::new`](https://ruby-doc.org/core-2.5.0/Thread.html#method-c-new) 🙃

Reply:
Now that you pointed that out, I think it sounds reasonable. Iirc, ruby passes everything by reference except immutable values (like integers), but just to be on the safe side we can move the declaration of `release_notes` above `upload_worker` and let the scope see it, as you suggested.

---

> what happens if multiple UploadJobs are finishing potentially out of order and try to append to that array concurrently?

I don't think the order of the results matters that much since in `upload_changelogs` there's nothing that would depend on it 👀